### DISCUSSION TITLE
RFC - Upgrade minimum .Net version

### DISCUSSION THREAD

**Discussion by mitchcapper** [cefsharp/CefSharp#4399] [General]
with .net 4.5.2 EOL being April of 2022 and VS installers no longer offering the package (although can still be installed separately) any plans to switch to 4.6.2 for the minimum version?  4.6.2 still works on W7.

**Comment by amaitland** [comment]
With [Google Chrome ending support for Windows Windows 7/8/8.1](https://support.google.com/chrome/thread/185534985/sunsetting-support-for-windows-7-8-1-in-early-2023?hl=en) the though had crossed my mind increase the minimum required version.

There would need to be a few things considered:

### How much notice to give? 3 mths, 6mths? 

We should give a notice period of at least a few months. There's a new major release every month, so that's roughly 3 or 6 versions (or somewhere in between).

### Which version? .Net 4.6, 4.6.1, 4.6.2?

`Windows 10` is the minimum required `OS`, it ships with the `.Net Framework` pre-installed. I believe every version of `Windows 10` has at least [.Net 4.6](https://learn.microsoft.com/en-us/dotnet/framework/migration-guide/versions-and-dependencies#net-framework-46) included by default. `.Net 4.6.2` comes installed by default with `10 Anniversary Update (Version 1607)`, personally that I think would be an acceptable minimum.

### How many users does dropping .Net .4.5.2 impact?

This is a tough one to measure. If we say choose `.Net 4.6` then if I understand correctly, it should be installed on every computer that `CefSharp` will run on, so anyone targeting an older version would only need to change their target version, they shouldn't need to worry about installing a newer runtime on client machines. 

Searching though [issues](https://github.com/search?q=repo%3Acefsharp%2FCefSharp+%224.5.2%22&type=issues&s=updated&o=desc)/[discussions](https://github.com/search?q=repo%3Acefsharp%2FCefSharp+%224.5.2%22&type=discussions&s=updated&o=desc) for `4.5.2`, there was a [discussion](https://github.com/cefsharp/CefSharp/discussions/4397) started just today using the framework. There hasn't been an issue created in a while, so from the limited data I have, the number is probably fairly small.

[.Net 4.6.2 Developer Pack](https://support.microsoft.com/en-us/topic/microsoft-net-framework-4-6-2-developer-pack-and-language-packs-5b1b177e-b44e-1e34-7741-1c10c2d62924#:~:text=The%20.-,NET%20Framework%204.6.,%2C%20or%20third%2Dparty%20IDEs.) has pretty decent support for older IDE's, `VS2012/2013/2015` so I think that's ok.

### What does dropping .Net 4.5.2 get us?

- Simplified `Visual Studio` installation (no need to install an outdated `SDK` separately).
- There are some newer API's which would allow for code cleanup/improvements

### Other considerations
- From an end user point of view it's the `minimum required` version, any newer version of `.Net` can be used. Using a newer version is highly recommended.
- Community backlash, when I made `.Net 4.5.2` the minimum, one case where I was borderline abused for not lowering the minimum. I'm not keep to repeat that personally.


Personally my choice would upgrading the minimum to `.Net 4.6.2` with a `3mth` notice period. Be great to hear feedback from the wider community.

**Comment by mitchcapper** [comment]
Well 4.6 was EOL in 2022 but 4.6.2 came out 7 years ago and is good for another 4 with the same OS support.   VS2012 came out over a decade ago and went EOL earlier this year.  The ability to still support that is quite the broad range.

While we all know its ridiculous for you to be berated about not providing enough free support and contributions, it is the internet.   Setting a standard may help, where the bare minimum version guaranteed to support is products that are not EOL and may be set higher if significant developer effort is required otherwise.    If someone needs something outside that scope they can:
*) Use their current version there are no forced upgrades
*) Maintain a branch with older dependency support themselves
*) Contact the cefsharp team and see about sponsoring work on an older version builds being made on demand one offs.
*) Look for alternative free solutions

Given the massive lifespan for 4.6.2 and the fact 4.7 came out some 9 months later it may be desired to move to that before the 4.6.2 EOL.  This given some of the modernization that 4.7 brings too might avoid some bug reports.

A better option may simply be committing to notice of X months prior to an upgrade as done above.   With chrome releases every month that 3 month window means for 3 chrome builds it can be at the top of the release the minimum .net version is changing.

**Reply by amaitland** [reply]
> Well 4.6 was EOL in 2022 but 4.6.2 came out 7 years ago and is good for another 4 with the same OS support.

`4.6.2` would be my personal preference. I would like to keep the impact on end users low, staying at `.Net 4.5.2` isn't a huge burden.

> Setting a standard may help, where the bare minimum version guaranteed to support is products that are not EOL

This is the approach I'd tried in the past. At the time of upgrade `.Net 4.5.2` was the lowest supported version. There are some who will disagree with this approach e.g. https://github.com/cefsharp/CefSharp/issues/1785#issuecomment-299068510

> A better option may simply be committing to notice of X months prior to an upgrade as done above. With chrome releases every month that 3 month window means for 3 chrome builds it can be at the top of the release the minimum .net version is changing.

`M114` is due in a little over 3 months time. Tentatively I think we'll look at upgrading to **.Net 4.6.2** in **M114**. 

I would like to hear from the boarder community, anyone who would be impacted by this change please leave a comment below.

**Comment by n303ef** [comment]
As long as we working with DOM we always have to use minimum 4.7.2 as CefSharp.DevTools.Dom requires.

**Comment by amaitland** [comment]
Is anyone still targeting `.Net 4.5.2`, `.Net 4.6` or `.Net 4.6.1` that would be unable to upgrade to `.Net 4.6.2`?

**Reply by aimcosoftware** [reply]
We're still using 4.5.2, but don't think there would be any issues moving to 4.6.2

**Reply by amaitland** [reply]
Thanks for confirming, please let me know if it looks like there will be any issues upgrading 👍

**Reply by aimcosoftware** [reply]
Just changed three of our Winforms CEFSharp projects from 4.5.2 to 4.6.2 with no issues.

**Comment by cwollenhaupt** [comment]
Most of the web servers and APIs that we interact with only support TLS 1.2 and 1.3 or announced that for the near future. Even though our customers use their computer as long as possible, we moved about everyone to at least Windows 7/Windows Server 2012. This makes .NET Framework 4.7.2 the minimum version that we support.

**Comment by amaitland** [comment]
The upgrade (#4482) will be delayed until `M115` to ensure everyone can easily upgrade to [Chromium 114.0.5735.110](https://chromereleases.googleblog.com/2023/06/stable-channel-update-for-desktop.html) which includes a fix for CVE-2023-3079. Google is aware that an exploit for CVE-2023-3079 exists in the wild.
